### ISSUE TITLE
HDMI/DP audio out not producing sound

### ISSUE THREAD

**Issue by lyncolnmd** [jmontleon/pixelbook-fedora#55]
First tims opening an issue so I hope I'm doing this right. Im glad audio is working but I think the last thing to figure out is why HDMI/DP audio out has not been producing sound on any distro thats not Ubuntu 22.04 LTS or its derivatives. Pulseaudio seems to be working fine. Unsure if there are missing plugins. pauvcontrol shows output graphically and not seeing any difference in things that are muted in alsamixer. Hoping we can figure this out. Its a long-standing issue that has gone unnoticed by most. Ubuntu must have some configuration/package that makes it work...

**Comment by lyncolnmd** [comment]
Thanks Jason. It specifically works on Ubuntu 22.04 LTS. Its the only distro that has everything working (considering brightness was fixed after Kernel 5.17 and the LTS just got 5.19 in a point release.) They were also on Pulseaudio out of the box so no need to remove pipewire. It also works on Mint if youd like to try that. My github has the sound drivers as deb files for a rapid install if you need to test.

**Comment by jmontleon** [comment]
Awesome, thanks for the details! That will definitely make it easier to get a working install going to compare.

**Comment by jmontleon** [comment]
Can you give the output from `uname -a` and `pacmd list-cards` to start. I have Ubuntu installed and HDMI also does not work. That's not all bad though. More like comparing Apples to Apples than Apples to Oranges. Maybe we can find out what is different between the two installs.

Also, what HDMI adapter are you using. All I have are DP and HDMI versions of this. I _think_ anything else I have is Thunderbolt so not going to work... https://www.amazon.com/gp/product/B01M3PTUV6/

**Comment by lyncolnmd** [comment]
Just as an FYI you have to try one output because the other is for DP audio out. I had edited my UCM file to label them differently. Ill double check shortly. Which version of Ubuntu are you on? Thats very odd that its not working on it considering it has been all along.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jmontleon** [comment]
I installed fromm the 22.04.2 LTS ISO image at https://ubuntu.com/download/desktop

**Comment by lyncolnmd** [comment]
Maybe they changed something in the second point release? I know it had been working on both Ubuntu and Linux Mint. Ill be able to provide you with the output a little later

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by lyncolnmd** [comment]
Hey Jason,
Here are the outputs your requested:
```
[david@archlinux ~]$ uname -a
Linux archlinux 6.2.2-arch1-1 #1 SMP PREEMPT_DYNAMIC Fri, 03 Mar 2023 15:58:31 +0000 x86_64 GNU/Linux
[david@archlinux ~]$ pacmd list-cards
1 card(s) available.
    index: 0
	name: <alsa_card.platform-kbl_r5514_5663_max>
	driver: <module-alsa-card.c>
	owner module: 6
	properties:
		alsa.card = "0"
		alsa.card_name = "kbl-r5514-5663-max"
		alsa.long_card_name = "Google-Eve-1.0"
		alsa.driver_name = "snd_soc_kbl_rt5663_rt5514_max98927"
		device.bus_path = "platform-kbl_r5514_5663_max"
		sysfs.path = "/devices/platform/kbl_r5514_5663_max/sound/card0"
		device.form_factor = "internal"
		device.string = "0"
		device.description = "Built-in Audio"
		module-udev-detect.discovered = "1"
		device.icon_name = "audio-card"
	profiles:
		HiFi: Default (priority 8000, available: unknown)
		off: Off (priority 0, available: unknown)
	active profile: <HiFi>
	sinks:
		alsa_output.platform-kbl_r5514_5663_max.HiFi__hw_kblr55145663max_7__sink/#0: Built-in Audio HDMI Audio
		alsa_output.platform-kbl_r5514_5663_max.HiFi__hw_kblr55145663max_6__sink/#1: Built-in Audio Display Port Audio
		alsa_output.platform-kbl_r5514_5663_max.HiFi__hw_kblr55145663max_2__sink/#2: Built-in Audio Headphones
		alsa_output.platform-kbl_r5514_5663_max.HiFi__hw_kblr55145663max_0__sink/#3: Built-in Audio Speaker
	sources:
		alsa_output.platform-kbl_r5514_5663_max.HiFi__hw_kblr55145663max_7__sink.monitor/#0: Monitor of Built-in Audio HDMI Audio
		alsa_output.platform-kbl_r5514_5663_max.HiFi__hw_kblr55145663max_6__sink.monitor/#1: Monitor of Built-in Audio Display Port Audio
		alsa_output.platform-kbl_r5514_5663_max.HiFi__hw_kblr55145663max_2__sink.monitor/#2: Monitor of Built-in Audio Headphones
		alsa_output.platform-kbl_r5514_5663_max.HiFi__hw_kblr55145663max_0__sink.monitor/#3: Monitor of Built-in Audio Speaker
		alsa_input.platform-kbl_r5514_5663_max.HiFi__hw_kblr55145663max_1__source/#4: Built-in Audio Headset Microphone
		alsa_input.platform-kbl_r5514_5663_max.HiFi__hw_kblr55145663max_4__source/#5: Built-in Audio Internal Microphone
	ports:
		[Out] HDMI2: HDMI Audio (priority 100, latency offset 0 usec, available: unknown)
			properties:
				
		[Out] HDMI1: Display Port Audio (priority 200, latency offset 0 usec, available: unknown)
			properties:
				device.product.name = "C27F398"
		[Out] Headphones: Headphones (priority 800, latency offset 0 usec, available: unknown)
			properties:
				
		[Out] Speaker: Speaker (priority 900, latency offset 0 usec, available: unknown)
			properties:
				
		[In] Mic: Headset Microphone (priority 100, latency offset 0 usec, available: unknown)
			properties:
				
		[In] Internal Mic: Internal Microphone (priority 100, latency offset 0 usec, available: unknown)
			properties:
				
[david@archlinux ~]$ 
```

**Comment by lyncolnmd** [comment]
Did you try it on Mint?
Some of the things I have been looking into are:

- Sample rates/engine differences that need to be configured
- The possibility of missing pulseaudio components
- Kernel related issues (Very unlikely as it didn't work with LTS on Arch back when LTS was stil 5.15)
- Possible missing alsa components/configurations

**Comment by jmontleon** [comment]
It's not working on Arch though, right? I was looking for the output from Ubuntu where you have/had it working. Is that possible, or is your install gone?

If not I can give it a try on Mint when I have a few to install.

**Comment by lyncolnmd** [comment]
Its not working on Arch. I dont have Ubuntu installed at the moment. But Mint does have it working for sure. Also based on pacmd-list cards im seeing the monitor so its detecting it as a source for output. I just dont know why i hear nothing

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jmontleon** [comment]
Is Mint 21.1 what you used?

**Comment by lyncolnmd** [comment]
Last time I checked it was working on Vera (21.1) installer debs are on my girhub for convenience

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jmontleon** [comment]
Vera 21.1 does work. It's because they're still on the 5.15 LTS kernel. I imagine it probably worked on 22.04 LTS until they jumped from 5.15 to 5.19.

It broke way back and no one noticed.

Running through kernel RPMs on Fedora 37 it worked with:
[kernel-5.17.0-0.rc0.20220112gitdaadb3bd0e8d.63.fc36](https://koji.fedoraproject.org/koji/buildinfo?buildID=1877179)

but not:
[kernel-5.17.0-0.rc2.83.fc36](https://koji.fedoraproject.org/koji/buildinfo?buildID=1910212) 

So about 12 builds to narrow it down:
```
$ git bisect good daadb3bd0e8d
status: waiting for bad commit, 1 good commit known
[jason@towerofpower linux]$ git bisect bad 26291c54e111ff6ba87a164d85d4a4e134b7315c
Bisecting: 3070 revisions left to test after this (roughly 12 steps)
[aee101d7b95a03078945681dd7f7ea5e4a1e7686] powerpc/64s: Mask SRR0 before checking against the masked NIP
```

**Comment by lyncolnmd** [comment]
Whoa Jason that was fast. I guess I didnt try anything prior to the linux lits package on Arch because in my mind it was still kernel 5.15 so it should have been enough for a comparison. Ive noticed it for quite some time now but i wasnt sure if it was a kernel level issue. If this gets fixed then this makes work on the pixelbook pretty complete. Otherwise Ive been tinkering with the UCM2 files to make using the sound configs a little smoother... Made some simple additions to improve the experience. Cant get pulse/alsa to autodetect the jack but im pretty sure an acpid script should be enough to get that working

**Comment by jmontleon** [comment]
Ya, there's an acpid script for the headphone jack in my repo. I was looking at the newer AVS driver, which is not ready for our sound card (https://github.com/eupnea-linux/audio-scripts/issues/39 has details) and noticed there was one there too.

Mine is just rolled in with the other acpi stuff:
https://github.com/jmontleon/pixelbook-fedora/blob/main/scripts/pixelbook-acpi#L38-L47
https://github.com/jmontleon/pixelbook-fedora/blob/main/configs/pixelbook-acpi.service

https://github.com/eupnea-linux/audio-scripts/tree/main/configs/audio/avs/auto-switcher

**Comment by lyncolnmd** [comment]
Ill definitely try to apply the scripts in my own setup im more concerned about the HDMI/DP out though especially because it makes my experience using a Docking station so much better. If theres anything you need from me to help you narrow down what went wrong with the kernel let me know (even though my technical knowledge isnt as extensive as yours)

**Comment by jmontleon** [comment]
From a bisect, `636110411ca726f19ef8e87b0be51bb9a4cdef06` is where it first stopped working. Reverting this on 5.17.14 makes it work again. It unfortunately doesn't revert cleanly anymore from 5.18 onward, and my attempt at fixing it up built, but didn't result in audio anymore. 

It didn't finish up until the early hours of this AM so I'll stare, do some searching, and likely email the linux regression list once I've exhausted other options and gathered what info I can to see if we can get a fix.

**Comment by jmontleon** [comment]
https://lore.kernel.org/regressions/CAJD_bPKQdtaExvVEKxhQ47G-ZXDA=k+gzhMJRHLBe=mysPnuKA@mail.gmail.com/T/#u

**Comment by lyncolnmd** [comment]
Thanks so much for this Jason. I hope it gets fixed soon as I frequently use audio output through my monitor. I think this is literally the last issue that needs fixing on the PB with linux.

**Comment by jmontleon** [comment]
You're welcome. I built a test kernel for Fedora here, but if I'm not mistaken you're using Arch? https://copr.fedorainfracloud.org/coprs/jmontleon/pixelbook-test/builds/

I don't know what your comfort level is with building custom kernels, but if you're able to test the patch and comment on the regression thread that you're affected and it's working it might speed things along. Sometimes more voices help. Otherwise if it goes a week or two with no activity I can respond or try to track someone down to look.

I've used this doc to build kernels on Arch:
https://wiki.archlinux.org/title/Kernel/Arch_Build_System#:~:text=The%20Arch%20Build%20System%20can,configuration%20or%20add%20additional%20patches.

I think the only place it falls down is regarding updating the bootloader, because there are plenty of options they don't detail any of them and that can without updating it you'll just keep booting your stock kernel. Basically you should see whatever you update pkgbase to in step 2 in your uname output. It's been awhile, but for grub if memory serves, the grub-mkconfig command from 3.1.1 is probably what you're after, for example. https://wiki.archlinux.org/title/GRUB#Generate_the_main_configuration_file

If you're interested in the nitty gritty of what broke:
`skl_link_hw_params` used to call `snd_soc_dai_set_tdm_slot` in `skl-pcm.c`. The conditional is based upon whether it's playback or input.
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/sound/soc/intel/skylake/skl-pcm.c?h=linux-5.16.y#n566

This function looked for `set_tdm_slot` on the `snd_soc_dai_ops` struct and ran the function specified.
With the offending commit `skl-pcm.c` now calls `snd_soc_dai_set_stream` which looks for `set_stream` instead.
```
diff --git a/sound/soc/intel/skylake/skl-pcm.c b/sound/soc/intel/skylake/skl-pcm.c
index 9ecaf6a1e8475..8378c187959fb 100644
--- a/[sound/soc/intel/skylake/skl-pcm.c](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/sound/soc/intel/skylake/skl-pcm.c?h=v6.3-rc2&id=e8444560b4d9302a511f0996f4cfdf85b628f4ca)
+++ b/[sound/soc/intel/skylake/skl-pcm.c](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/sound/soc/intel/skylake/skl-pcm.c?h=v6.3-rc2&id=636110411ca726f19ef8e87b0be51bb9a4cdef06)
@@ -562,11 +562,8 @@ static int skl_link_hw_params(struct snd_pcm_substream *substream,
 
 	stream_tag = hdac_stream(link_dev)->stream_tag;
 
-	/* set the stream tag in the codec dai dma params  */
-	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
-		snd_soc_dai_set_tdm_slot(codec_dai, stream_tag, 0, 0, 0);
-	else
-		snd_soc_dai_set_tdm_slot(codec_dai, 0, stream_tag, 0, 0);
+	/* set the hdac_stream in the codec dai */
+	snd_soc_dai_set_stream(codec_dai, hdac_stream(link_dev), substream->stream);
 
 	p_params.s_fmt = snd_pcm_format_width(params_format(params));
 	p_params.ch = params_channels(params);
```

`snd_soc_dai_set_stream` and `set_stream` take slightly different input, so the rest of my [PR](https://gist.github.com/jmontleon/4780154c309f956d97ca9a304a00da3f) is just adjusting for that.

`stream_tag` and `hdac_stream(link_dev)->stream_tag` are the same. It's just being received as unstructured data and needs to be cast back into an hdac_stream (_I think_ my C skills are minimal; and I don't come from a strong development background, mostly just roll my face on my keyboard and get good outcomes now and then).

Basically, I think hdac_hdmi just got left behind by accident.

**Comment by saxa** [comment]
Hi I will try this on Slackware too and on my fedora 37. Just dont want to steal the ticket, but on my fedora install I still have no audio with the latest updates. It shows all as working it have a sound card but I can not play any video with audio. Videos work if I mute the audio they start playing, otherwise they just stay there.
[pacmd_list_cards.txt](https://github.com/jmontleon/pixelbook-fedora/files/11001417/pacmd_list_cards.txt)

**Comment by lyncolnmd** [comment]
Hey Jason,
Thanks alot for your work... I am 100% the furthest thing from an expert. I just read alot and I have "the knack" (that thing where you just kinda understand how things work if even vaguely) so I should be able to build a custom kernel to see if it patches the error on arch. however it might take me some time to do it. ive been super super swamped. Still wanna try it though to make sure it works.
Currently I'm on kernel 6.2.6
while we work on this though, did you ever find out what causes the static noise and slow (like 0.5x speed) playback with pipewire?
Ill still be tinkering a bit with the UCM file to get the best UCM configuration I can find

**Comment by lyncolnmd** [comment]
> Hi I will try this on Slackware too and on my fedora 37. Just dont want to steal the ticket, but on my fedora install I still have no audio with the latest updates. It shows all as working it have a sound card but I can not play any video with audio. Videos work if I mute the audio they start playing, otherwise they just stay there. [pacmd_list_cards.txt](https://github.com/jmontleon/pixelbook-fedora/files/11001417/pacmd_list_cards.txt)

Hey saxa, no expert here but in my experience sometimes the installs just dont seem to pickup the audio. I find this particulatly true when using tools like an ansible playbook script to copy the files... Ironically, if the files are manually copied using root copy commands (i use the commands that preserve the permissions) then i find that it works every time.

**Comment by saxa** [comment]
@lyncolnmd I see, but where should I start ? I can copy files manually around, but I think that the scripts that build the packages I install on my fedora are well tested and working. The audio problem exists only on my fedora 37 install. My slackware install works perfectly well. I still see those -----[cut here]------ errors in the dmesg. I suspect this is related to the audio problems although I have similar ----[cut here]----- in slackwares dmesg too.

**Comment by lyncolnmd** [comment]
I used the same files pulled from the image using kpartx. I think I was having the same issue with Fedora once i started using the ansible on my github but i know for sure that copying the drivers manually helped... so i made a folder of the files so i didn't have to pull them everytime and manually copied under superuser. this is also why i kinda wanted to just make an rpm that could move the files where I wanted them but i was onl successful in making the deb.

**Comment by lyncolnmd** [comment]
@saxa and @jmontleon did we ever get our hardware working with pipewire? or make any headway in figuring out why they the audio hardware and pipewire dont play well together?

**Comment by saxa** [comment]
I am quite new to this pixelbook thingy, but as far as I know no, we havent had any pipewire working.
I tried it at the first install of fedora and audio have not worked. I could hear some sound but was all cut and sluggish. Unhearable. Tried it also in Slackware, same thing, I think even worst, no  audio at all iirc.

**Comment by jmontleon** [comment]
re: pipewire, no I submitted an issue and didn't even get a response. https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/1452

**Comment by jmontleon** [comment]
I was asked to post the fix as a patch to the list. Awaiting feedback
https://lore.kernel.org/regressions/20230324170711.2526-1-jmontleo@redhat.com/T/#u

**Comment by jmontleon** [comment]
Looks like that can probably go in. I'll leave this open until we have a stable kernel with it included, but there's probably not much left to do other than wait.

**Comment by jmontleon** [comment]
In linux-next https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/sound?id=f6887a71bdd2f0dcba9b8180dd2223cfa8637e85

**Comment by lyncolnmd** [comment]
Woor!!!! Way to Go Jason! Thanks for all you do for us PB users... I should probably see it show up in the upcoming Arch kernels in the coming weeks. Ill definitely let you know if I see it

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jmontleon** [comment]
Cool. It's in Mark Brown's branch for 6.3 so probably in the next (or at least one of the upcoming 6.3 RCs).
https://git.kernel.org/pub/scm/linux/kernel/git/broonie/sound.git/log/?h=for-6.3

My understanding is that since I included the `Cc: stable@...` tag on the patch it'll go to stable after it lands in Linus' tree. 

So with a little luck within a few 6.2 releases, hopefully.

**Comment by jmontleon** [comment]
It's in Linus' tree now.
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/sound?id=f6887a71bdd2f0dcba9b8180dd2223cfa8637e85

From this point I believe it can be added to stable, so hopefully in the next 6.2.11 or soon after.

**Comment by jmontleon** [comment]
In 6.3-rc6 now.

**Comment by jmontleon** [comment]
Tested with a Fedora kernel 6.3-rc6 build and HDMI works fine.
https://koji.fedoraproject.org/koji/buildinfo?buildID=2183600

**Comment by jmontleon** [comment]
Now in the following. To find in the log by search hdac_hdmi.
- 6.2.11 https://cdn.kernel.org/pub/linux/kernel/v6.x/ChangeLog-6.2.11
- 6.1.24 https://cdn.kernel.org/pub/linux/kernel/v6.x/ChangeLog-6.1.24
- 5.15.107 https://cdn.kernel.org/pub/linux/kernel/v5.x/ChangeLog-5.15.107

I believe it's due to go in the 5.10 LTS, probably 5.10.178. I haven't seen a release since I got the email for it going in the stable queues. Not sure anyone around here was ever using 5.10. I started here with Fedora 34, which was launched with 5.11 and may have been even newer. The earliest reference I saw in this repo was 5.13.8, so this one probably doesn't matter for anything.

**Comment by jmontleon** [comment]
Packages for
- f38 https://koji.fedoraproject.org/koji/buildinfo?buildID=2186187
- f37 https://koji.fedoraproject.org/koji/buildinfo?buildID=2186188
- f36 https://koji.fedoraproject.org/koji/buildinfo?buildID=2186189

I gave the F38 packages a try and HDMI works fine now.

**Comment by lyncolnmd** [comment]
Thanks Jason,
I got the 6.2.11 update for Arch and it works perfectly.

**Comment by jmontleon** [comment]
@lyncolnmd glad to hear it's working! I don't know if you saw, but we finally learned how to get pipewire working if that's something you're after as well. I've update the instructions in the README, playbook, etc. based on what's been learned.

**Comment by lyncolnmd** [comment]
Hey Jason. Pipewire is working great. No issues so far. THANK YOU!
Really enjoying a fully working Arch Linux
